### PR TITLE
[JIT] X64 Fix - Handle 'imul' instruction variants to determine if a register was written to

### LINKED ISSUE #85602: [Fuzzlyn] Failure for complex expression
This is new to .NET 8.

```csharp
// Generated by Fuzzlyn v1.5 on 2023-05-01 15:58:26
// Run on X64 Linux
// Seed: 15325437049811320331
// Reduced from 7.4 KiB to 1.0 KiB in 00:00:28
// Debug: Outputs 18446744069414922151
// Release: Outputs 337831
public class C0
{
    public C0(ulong f0, sbyte f1, byte f2, ulong f3, ulong f4, int f5, short f6)
    {
    }
}

public class Program
{
    public static ushort s_1;
    public static void Main()
    {
        M0();
    }

    public static void M0()
    {
        ulong var0 = ~(ulong)(uint)(-15356 * (2637600427U % (byte)((2147483647 ^ (4167361218894137384UL * (ushort)(-(ushort)(-(ushort)(128 * (5114990800133743712L % (byte)((byte)(-(byte)(-(byte)(-(byte)~(byte)(-2 % (short)((short)~(short)(-(short)(-90400400 - (-(0 | ~~(int)(32766 & (uint)(-~(uint)(-(uint)(629572031969723397L ^ (sbyte)(1UL * (byte)(8945663325738713761L / ((-(long)(127 / (sbyte)((sbyte)(-23817 % (ushort)((ushort)~M1(new C0(8144643251292930980UL, -118, 183, 18446744073709551614UL, 1827525571111008345UL, 1751590714, -32653)) | 1)) | 1))) | 1))))))))))) | 1))))) | 1))))))) | 1)));
        System.Console.WriteLine(var0);
    }

    public static ushort M1(C0 argThis)
    {
        return s_1;
    }
}
```

### PR DESCRIPTION
**Pull request by TIHan**
Resolves https://github.com/dotnet/runtime/issues/85602

There were `imul` variants that refer to which register is implicitly being written to. We needed to handle these cases for `emitIsInstrWritingToReg`.

Local diffs - Windows x64:
**-60 bytes** improvement
No regressions.

### CODE REVIEW

File: src/coreclr/jit/emitxarch.cpp [hunk 1]
```diff
@@ -662,6 +662,120 @@ bool emitter::emitIsInstrWritingToReg(instrDesc* id, regNumber reg)
         case INS_call:
             return true;
 
+        case INS_imul_AX:
+            if (reg == REG_RAX)
+            {
+                return true;
+            }
+            break;
```
Comment (tannergooding) [inline src/coreclr/jit/emitxarch.cpp:670]: nit: Group these all as one case and use `if (reg == inst3opImulReg(ins))` instead

Comment (TIHan) [reply src/coreclr/jit/emitxarch.cpp]: done

**Review by tannergooding — APPROVED** [review]

**Comment by TIHan** [comment]
@dotnet/jit-contrib ready - just need a sign-off - [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=258492&view=ms.vss-build-web.run-extensions-tab)

**Review by jakobbotsch — APPROVED** [review]
Why does it result in improvements?

**Comment by TIHan** [comment]
It caught a few more cases since we accurately know if a register was being written to for imul.

**Pull request merged**